### DISCUSSION TITLE
Allow reusable workflows to be located at arbitrary locations and be local

### DISCUSSION THREAD

**Discussion by umarcor** [community/community#9050] [Actions]
Although the documentation does not specify the limitation, it seems that reusable workflows must be located in `.github/workflows`. Moreover, specifying the version is necessary. Therefore, it is not possible to have a reusable workflow in the current branch.

See 

> https://docs.github.com/en/actions/learn-github-actions/reusing-workflows#calling-a-reusable-workflow
> You reference reusable workflow files using the syntax:
>
> {owner}/{repo}/{path}/{filename}@{ref}

and 

https://github.com/hdl/containers/actions/runs/1469619022
> invalid value workflow reference: references to workflows must be rooted in '.github/workflows'

and

https://github.com/hdl/containers/actions/runs/1469617309
> invalid value workflow reference: no version specified

It would be desirable if the syntax was changed to `[{owner}/{repo}/]{path}/{filename}.yml[@{ref}]`.

Refs:

- actions/runner#1493
- https://github.community/t/allow-reusable-workflows-to-be-located-at-arbitrary-locations-and-be-local/212745
- https://github.community/t/ref-head-in-reusable-workflows/203690/10
- #8311
- https://github.com/github/feedback/discussions/7835#discussioncomment-1843575

**Comment by PathogenDavid** [comment]
In [another community thread](https://github.community/t/ref-head-in-reusable-workflows/203690/10), @jenschelkopf shared that they're working on something similar to this:

> Hi everyone. I’m the product manager for reusable workflows. Today, you have to provide the full, static path and ref for the reusable workflow. We don’t evaluate any expressions.
>
> We do plan to support this at some point, and I’ll update this thread when we do.

And later in the thread:

> [...]
> It will likely look a lot like referencing a local action, but without the need to do a checkout first
> [...]

As far as I'm aware there's no public roadmap item for this yet.

**Comment by jenschelkopf** [comment]
Hi @umarcor and @PathogenDavid, we just released the ability to reference a local reusable workflow: https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally

**Reply by PathogenDavid** [reply]
Thanks a lot for getting this feature added! Just tried it out in my test repo and it works great!

**Reply by umarcor** [reply]
@jenschelkopf is `{path}` still hardcoded to `.github/workflows`?

**Comment by Hazmi35** [comment]
Hello, sorry for the bump, any updates regarding this?

**Reply by PathogenDavid** [reply]
This has been completed since January, see [this blog post](https://github.blog/changelog/2022-01-25-github-actions-reusable-workflows-can-be-referenced-locally/) for details.

**Reply by umarcor** [reply]
@PathogenDavid the title of this discussion starts with "Allow reusable workflows to be located at arbitrary locations". Do you know whether that's supported already? Ref: https://github.com/orgs/github-community/discussions/9050#discussioncomment-2167487

**Reply by PathogenDavid** [reply]
@umarcor My bad, can confirm that workflow files still have to be in the `.github/workflows` folder. I wasn't paying enough attention to the specific thread I was in.

**Reply by jdogmcsteezy** [reply]
Still looking forward to being able to reference a workflow from and path.  It would allow us to reference submodule workflows. Our current workaround (because they are public repos) is to reference the workflow using [{owner}/{repo}//{filename}.yml[@{ref}] but needing to specify the ref can be quite tedious when making changes to the submodule workflow

**Comment by o-kasian** [comment]
Is this available in GitHub Enterprise, and if so - what version?

**Reply by aciokler** [reply]
Doesn't seem like it is.

**Comment by moritzluedtke** [comment]
I can confirm that as of today we still need to put reusable workflows into `.github/workflows` and nowhere else. I don't see why it was implemented that way, because we already need to give the exact path to the workflow file in the jobs `uses` segment. An explanation from GitHub would be much appreciated.

**Reply by cortwave** [reply]
Totally agree, if you have a big repo and a big team uses it then workflow dir becomes just a trash bin and it will be cool to have the possibility to structure it

**Reply by vladimirevstratov** [reply]
it will be a great feature!

**Reply by doubleforte** [reply]
We are also hoping for a way to structure our workflow dir so we could have something like:

```
├── .github
│   └── workflows
│       ├── main.yml
│       └── shared
│           ├── reuse-this-elsewhere.yml
│           └── reuse-this-one-too.yml
```

... or even better, not require them to be in the `.github` or `workflows` dir at all ...

```
├── reusable-workflow
│   ├── reuse-this-elsewhere.yml
│   └── reuse-this-one-too.yml
```

**Reply by mnadher** [reply]
Do they have any plan for implementing this in future ?

**Reply by gabrielsclimaco** [reply]
Any news on this?

**Reply by aciokler** [reply]
Much needed feature. Please implement. thanks

**Comment by romeuhcf** [comment]
It would be great if we could do something like:

```yaml
jobs:
  build-images:
    strategy:
      matrix:
        target: [frontend, backend]
    uses: myorg/myproj-${{ matrix.target }}/.github/workflows/build.yml@main
```

**Reply by umarcor** [reply]
@romeuhcf I believe that's a different feature request. It's about allowing variables in the `uses` field, not about the arbitrary location of workflows in the repos. See https://github.com/orgs/community/discussions/9049 instead.

Ref: https://github.com/hdl/containers/issues/48

**Comment by elijaholmos** [comment]
Well, I stumbled upon a seemingly stupid restriction: **reusable workflows, if located in an external repository, cannot be located in a repository called ".github"**. I say this is stupid, because it would seem natural to place code common to all repositories within an org/user inside the special .github repository. 

The following error is yieled by GH actions when I attempt to place a reusable workflow in my elijaholmos/.github repo:
![image](https://github.com/community/community/assets/35435704/70b23c0e-184e-474f-8e85-de978f7906b5)

However, moving the workflow to a separate repository with any other name (elijaholmos/elijaholmos in this case), the action runs without problems:
![image](https://github.com/community/community/assets/35435704/2c35dda4-de4e-4a27-8b76-a19bf58c92d0)

Both repos as well as their actions logs are public, so feel free to peruse for more information.

**Reply by jcbhmr** [reply]
actually it's because the path there is:

```
elijaholmos/.github/workflows/pull-request-title.yml@main
👇
elijaholmos/.github | workflows/pull-request-title.yml | main
```
which is NOT a repo with a .github/workflows/* yml file.

instead, just make sure that you do this:

```
elijaholmos/.github | .github/workflows/pull-request-title.yml | main
👇
elijaholmos/.github/.github/workflows/pull-request-title.yml@main
```

which _should_ work

**Reply by elijaholmos** [reply]
alas, @jcbhmr, this _does_ work- thank you for your help! The redundancy is still dumb, although I suppose that is included in the above discussions.

**Reply by umarcor** [reply]
@elijaholmos that's precisely what "to be located at arbitrary locations" means in the title :wink:.

**Comment by n3dst4** [comment]
My use case is that I have a subrepo (like a submodule but better) for a lot of shared code. I would like to be able to reference reusable workflows by arbitrary relative path to get the version that's in the shared subrepo. At the moment I have to use the repo link and remember to update the `{ref}` when needed.

**Comment by TechnicallyJoe** [comment]
We're moving from Azure DevOps to GitHub. In ADO we have an extensive library of pipeline templates that you are free to organize however you want. The limitation that your reusable workflow has to be in .github/workflows seems very odd.

Any news on this?

**Comment by PaulRBerg** [comment]
This limitation makes it difficult to use GitHub Actions to test GitHub Actions itself.

We have a repo with reusable workflows that has grown big over time, and the workflows are currently defined under `.github/workflows`.

Now, we would like to add some CI checks for the reusable workflows. While we can do this, it's a bit misleading to put the production code and the test code under the same roof.

Any ETA for when this basic functionality will be supported?

**Comment by borhara** [comment]
Any news on this?
